### ISSUE TITLE
.NET Core Logging only logs for the first test case

### ISSUE THREAD

**Issue by jcurl** [nunit/nunit#3919]
I'm attaching a very simple example on reproduction [testlog.zip](https://github.com/nunit/nunit/files/6988275/testlog.zip)

If I run `TestCase1` on its own, or `TestCase2` on its own, then output from logging is captured and can be analysed. If I run them both, then generally the output of `TestCase1` is captured, but not `TestCase2`, depending on which test case runs first (e.g. `TestCase1`). If I modify sources so that the `ILoggerFactory` is instantiated before every test case, then it appears to work as expected, but this makes the assumption that code under test won't do this (and so this is not a suitable workaround).

I used dotPeek as a symbol/source server, so some of the code might differ from what's published as sources. I was able to determine that the problem comes down to how the logs are written into a stringbuilder, but I do not understand enough of the NUnit sources to figure out where the root cause is, only that the underlying code is getting the log messages, but not recording it to the output.

The internal class `Microsoft.Extensions.Logging.Logger.Log<TState>(LogLevel, EventId, TState, Exception, Func<TState, Exception, string>)` is called.

```csharp
    public void Log<TState>(
      LogLevel logLevel,
      EventId eventId,
      TState state,
      Exception exception,
      Func<TState, Exception, string> formatter)
    {
      MessageLogger[] messageLoggers = this.MessageLoggers;
      if (messageLoggers == null)
        return;
      List<Exception> exceptions = (List<Exception>) null;
      for (int index = 0; index < messageLoggers.Length; ++index)
      {
        ref MessageLogger local = ref messageLoggers[index];
        if (local.IsEnabled(logLevel))
          LoggerLog(logLevel, eventId, local.Logger, exception, formatter, ref exceptions, in state);
      }
      if (exceptions == null || exceptions.Count <= 0)
        return;
      Logger.ThrowLoggingError(exceptions);
```

The value `local.IsEnabled(Debug)` is `true`, and `LoggerLog` is called, which just calls:

```csharp
logger.Log<TState>(logLevel, eventId, state, exception, formatter);
```

The above goes into `ConsoleLogger.Log<TState>(...)`, it writes successfully the `WriteMessage` internally, which adds to a queue, and a separate thread is then responsible for writing this to the output. This is done by 

```csharp
this._queueProcessor.EnqueueMessage(message1);
```

Hence we can tell that the Microsoft Logger did receive the log message.

Then method `EnqueueMessage` is in the class `ConsoleLoggerProcessor`, and internally puts the message on a queue.

```csharp
this._messageQueue.Add(message);
return;
```

In parallel a thread is reading from the queue and writing to the console calling `WriteMessage`.

```csharp
foreach (LogMessageEntry consuming in this._messageQueue.GetConsumingEnumerable())
  this.WriteMessage(consuming);
```

In the `WriteMessage` we see:

```csharp
internal virtual void WriteMessage(LogMessageEntry message) {
  IConsole console = message.LogAsError ? this.ErrorConsole : this.Console;
  if (message.TimeStamp != null)
    console.Write(message.TimeStamp, message.MessageColor, message.MessageColor);
  if (message.LevelString != null)
    console.Write(message.LevelString, message.LevelBackground, message.LevelForeground);
  console.Write(message.Message, message.MessageColor, message.MessageColor);
  console.Flush();
}
```

and that the `console` is a `Sytem.IO.TextWriter.SyncTextWriter`, and it has an internal property `_out` which is of type `NUnit.Framework.Internal.Execution.TextCapture`.

It is simply defined in `WindowsLogConsole` as:

```csharp
this._textWriter = stdErr ? System.Console.Error : System.Console.Out;
```

I put a breakpoint in `NUnit.Framework.Internal.Execution.TextCapture.Write` here:

```csharp
    public override void Write(string value)
    {
      TestExecutionContext currentContext = TestExecutionContext.CurrentContext;
      if (currentContext != null && currentContext.CurrentResult != null)
        currentContext.CurrentResult.OutWriter.Write(value);
      else
        this._defaultWriter.Write(value);
    }
```

and I can see that `value` is the debug message I'm printing. For some reason though, it is not showing up in the output of my test case.

The actual `Write` method looks like:

```csharp
    public override void Write(string? value)
    {
      if (!this._isOpen)
        throw new ObjectDisposedException((string) null, SR.ObjectDisposed_WriterClosed);
      if (value == null)
        return;
      this._sb.Append(value);
    }
```

and the value is being appended to `_sb`, but appears as though it might be for the previous test case still, even though it isn't printed there.

So NUnit did receive the debug message, but there seems to be some synchronization missing that it doesn't show up in the debug outputs.

I'd like to know if this is a bug in NUnit, and if not, how it might be possible to workaround.

**Comment by Dreamescaper** [comment]
Logging to console is not a reliable way to log something to test results, especially when multiple threads are involved (see [here](https://docs.nunit.org/articles/nunit/technical-notes/nunit-internals/specs/Text-Output-from-Tests-Spec.html)). The recommended way is to use `TestContext.WriteLine` method.
I don't think there's any existing TestContext logging provider, but you can probably implement your own. I've modified [XUnitLoggerProvider](https://github.com/dotnet/aspnetcore/blob/main/src/Testing/src/Logging/XunitLoggerProvider.cs), you can try whether it works for you. You can add it via `.AddProvider(new NUnitLoggerProvider())`.

[NUnitLoggerProvider.cs](https://github.com/nunit/nunit/files/7035329/NUnitLoggerProvider.txt)

Concerning behavior, that you've observed. `TestExecutionContext.CurrentContext` property is similar to ThreadLocal variable, i.e. value depends on a thread. But with ConsoleLogger writing to console actually happens from single thread - the one processing 
the queue. I assume that `CurrentContext` always corresponds to the test which created the logger because of that.
